File: php/classes/PML/Tags/TagFunction.php
<?php namespace PML\Tags;

use PML\Parser\PMLParser;

class TagFunction extends Tag {

    public function output()
    {
        $atts = $this->getAttributes();
        $name = $atts['name'];
        $args = $this->getChildTagsByName('argument');


        echo '<?php function '. $name .'(' . $this->buildArgumentString($args) . ') { ?>';
        PMLParser::parse([]);
        echo '<?php } ?>';
    }

    /**
     * Build the arguments string
     *
     * @param Array $args
     */
    private function buildArgumentString($args)
    {
        $text = '';

        if(!empty($args)) {
            foreach($args as $arg) {
                $attrs = $arg->getAttributes();
                $name = $attrs['name'];
                $text .= $name . ', ';
            }
        }
        
        return substr(trim($text), 0, -1);
    }
}

File: java/PML/src/tags/TagFunction.java
package tags;

import java.util.ArrayList;

import parser.PMLParser;

public class TagFunction extends Tag{

	@Override
	public void output() {
		String functionName = this.getAttributes().get("name");
		String args = this.buildArgumentString(this.getChildTagsByName("argument"));
		
		System.out.println("<?php function " + functionName + "(" + args + ") { ?>");
		PMLParser.parse(this.getChildTagsByName("body"));
		System.out.println("<?php } ?>");
	}
	
	private String buildArgumentString(ArrayList<Tag> argTags) {
		String argString = "";
		
		for (Tag argTag: argTags) {
			argString += argTag.getAttributes().get("name") + ", ";
		}
		
		// Remove last two characters ", " and return.
		return argString.substring(0, argString.length()-2);
	}
}
